packages/Kokkos/test/tstUserApplication.cpp

```diff
@@ -0,0 +1,1403 @@
+// TEST TEMPLATE DECLARATIONS
+//---------------------------------------------------------------------------//
+TEUCHOS_UNIT_TEST_TEMPLATE_2_DECL( UserApplication, node_list, SC, NO )
+{
+    // Test types.
+    using DeviceType = typename NO::device_type;
+    using ExecutionSpace = typename DeviceType::execution_space;
+    using Scalar = SC;
+    using ParallelModel =
+        typename UserAppTest::GetParallelModel<ExecutionSpace>::Type;
+
+    // Create the test class.
+    const int space_dim = 3;
+    const int size_1 = 100;
+    const int size_2 = 5;
+    const int offset = 8;
+    const SC init_val = Teuchos::ScalarTraits<SC>::one();
+    auto user_test_class =
+        std::make_shared<UserAppTest::UserTestClass<Scalar, ExecutionSpace>>(
+            space_dim, size_1, size_2, offset, init_val );
+    auto void_ptr = std::static_pointer_cast<void>( user_test_class );
+
+    // Register the user functions.
+    auto registry =
+        std::make_shared<DataTransferKit::UserFunctionRegistry<Scalar>>();
+    registry->registerNodeListSizeFunction(
+        UserAppTest::nodeListSize<Scalar, ExecutionSpace>, void_ptr );
+    registry->registerNodeListDataFunction(
+        UserAppTest::nodeListData<Scalar, ExecutionSpace>, void_ptr );
+
+    // Create the user application.
+    DataTransferKit::UserApplication<Scalar, ParallelModel> user_app(
+        registry );
+
+    // Get a node list.
+    auto node_list = user_app.nodeList();
+
+    // Check the node list.
+    auto host_coordinates = Kokkos::create_mirror_view( node_list.coordinates );
+    Kokkos::deep_copy( host_coordinates, node_list.coordinates );
+    auto host_is_ghost_node =
+        Kokkos::create_mirror_view( node_list.is_ghost_node );
+    Kokkos::deep_copy( host_is_ghost_node, node_list.is_ghost_node );
+    for ( int i = 0; i < size_1; ++i )
+    {
+        for ( int d = 0; d < space_dim; ++d )
+            TEST_EQUALITY( host_coordinates( i, d ), i + d + offset );
+        TEST_EQUALITY( host_is_ghost_node( i ), 1 );
+    }
+}
+
+//---------------------------------------------------------------------------//
+TEUCHOS_UNIT_TEST_TEMPLATE_2_DECL( UserApplication, bounding_volume_list, SC,
+                                   NO )
+{
+    // Test types.
+    using DeviceType = typename NO::device_type;
+    using ExecutionSpace = typename DeviceType::execution_space;
+    using Scalar = SC;
+    using ParallelModel =
+        typename UserAppTest::GetParallelModel<ExecutionSpace>::Type;
+
+    // Create the test class.
+    const int space_dim = 3;
+    const int size_1 = 100;
+    const int size_2 = 5;
+    const int offset = 8;
+    const SC init_val = Teuchos::ScalarTraits<SC>::one();
+    auto user_test_class =
+        std::make_shared<UserAppTest::UserTestClass<Scalar, ExecutionSpace>>(
+            space_dim, size_1, size_2, offset, init_val );
+    auto void_ptr = std::static_pointer_cast<void>( user_test_class );
```

Comment:
This step is not necessary, you can do the `make_shared` directly on the void pointer.